src/Language/PureScript/Scheme/CodeGen/Optimizer.hs

```diff
@@ -1,28 +1,31 @@
 module Language.PureScript.Scheme.CodeGen.Optimizer where
 
-import Language.PureScript.Scheme.CodeGen.AST (AST(..), everywhere)
+import Language.PureScript.Scheme.CodeGen.SExpr (SExpr(..), everywhere)
+import Language.PureScript.Scheme.CodeGen.Scheme (app)
 
 
-optimizations :: AST -> AST
+optimizations :: SExpr -> SExpr
 optimizations = (simplifyLogic . specializeOperators)
 
 
 -- Single pass optimizer.
--- Run through all the AST expressions and apply the optimizations.
-runOptimizations :: [AST] -> [AST]
+-- Run through all the SExpr expressions and apply the optimizations.
+runOptimizations :: [SExpr] -> [SExpr]
 runOptimizations xs = map (\x -> everywhere optimizations x) xs
 
 
 -- TODO: possibly nasty hack. To review once externs are implemented.
-specializeOperators :: AST -> AST
+specializeOperators :: SExpr -> SExpr
 
-specializeOperators (Application (Application (Application (Identifier "Data.Semiring.add")
-                                 [Identifier "Data.Semiring.semiringInt"]) [x]) [y])
-  = Application (Identifier "+") [x, y]
+specializeOperators (List [List [List [Symbol "Data.Semiring.add",
+                                       Symbol "Data.Semiring.semiringInt"],
+                           x], y])
+  = app "+" [x, y]
 
```

Comment:
My experience when working with strongly typed languages is that "more intermediate representations is always better".
This is because "reusing types" is dangerous in a similar way to "reusing code" in that it risks being a poor fit for all the usecases involved where this doesn't happen with specialized code/types.
Another issue with having too few representations is that one risks having abrupt conversions, making the implementation harder to debug.

In this case we might be hitting both issues: doing optimizations on a `AST` that has `Application` is definitely less error prone than working on something that has just a `List` of stuff, and even if this doesn't seem clear cut here, future-you might agree.

So instead of replacing the previous `AST` type with `S-Expr` we might want to have both instead!
In this way:
- we could define optimizations on the more well-typed `AST`
- have a straightforward conversion from `AST` to `S-Expr`
- define the pretty printing only on `S-Expr`